### PR TITLE
KVL-242 Test for OUT_OF_RANGE errors

### LINKED ISSUE #7038: Test participant server support for Ledger API endpoint fail-over
# Context

The guidance for DAML application authors provided in #6875 needs to be tested. Due to time-pressure that has not been handled as part of that merge request yet.

# Acceptance Criteria

There is a test demonstrating a DAML App running against a Ledger with two API endpoints that serve the same Ledger API content. The test MUST

- [ ] exercise the support for command deduplication across the two endpoints
- [ ] exercise the support for `OUT_OF_RANGE` errors due to the secondary endpoint being behind.

FYI: @rautenrieth-da @gerolf-da

### PR DESCRIPTION
**Pull request by rautenrieth-da**
... when subscribing past the ledger end

Fixes #7038

### CODE REVIEW

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala [hunk 1]
```diff
@@ -140,6 +140,19 @@ private[testtool] final class ParticipantTestContext private[participant] (
   def currentEnd(overrideLedgerId: String): Future[LedgerOffset] =
     services.transaction.getLedgerEnd(new GetLedgerEndRequest(overrideLedgerId)).map(_.getOffset)
 
+  /**
+    * Gets the absolute offset that is beyond the current ledger end.
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:144]:
```suggestion
    * Gets an absolute offset that is beyond the current ledger end.
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala [hunk 2]
```diff
@@ -140,6 +140,19 @@ private[testtool] final class ParticipantTestContext private[participant] (
...
+    *
+    * Note: offsets are opaque byte strings, but they are lexicographically sortable.
+    * Prepending the current absolute ledger end with non-zero bytes creates an offset that
+    * is be beyond the current ledger end for the ledger API server.
+    * The offset might however not be valid for the underlying ledger.
+    * This method can therefore only be used for offsets that are only interpreted by the
+    * ledger API server and not sent to the ledger.
+    */
+  def futureOffset(): Future[LedgerOffset] =
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:153]: Can we rename this to something a but more deadly?

Perhaps "offsetWayPastLedgerEnd"?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionServiceIT.scala [hunk 3]
```diff
@@ -94,6 +94,40 @@ class TransactionServiceIT extends LedgerTestSuite {
       }
   })
 
+  test(
+    "TXAfterEnd",
+    "An OUT_OF_RANGE error should be returned for subscribing past the ledger end",
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionServiceIT.scala:99]:
```suggestion
    "An OUT_OF_RANGE error should be returned when subscribing past the ledger end",
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionServiceIT.scala [hunk 4]
```diff
@@ -94,6 +94,40 @@ class TransactionServiceIT extends LedgerTestSuite {
...
+        beyondEnd = request.update(_.begin := futureOffset, _.optionalEnd := None)
+        failure <- ledger.flatTransactions(beyondEnd).failed
+      } yield {
+        assertGrpcError(failure, Status.Code.OUT_OF_RANGE, "is after ledger end")
+      }
+  })
+
+  test(
+    "TXTreesAfterEnd",
+    "An OUT_OF_RANGE error should be returned for subscribing to trees past the ledger end",
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionServiceIT.scala:116]:
```suggestion
    "An OUT_OF_RANGE error should be returned when subscribing to trees past the ledger end",
```

**Review by SamirTalwar-DA — APPROVED** [review]

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala (hunk 2 above)
Comment (rautenrieth-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala]: Renamed to `offsetBeyondLedgerEnd`

**Pull request merged**